### PR TITLE
kmodule: support xz-compressed binaries

### PR DESCRIPTION
**Pull request by rminnich**
There are now several different ways to load modules, and there is a need
for xz-compressed modules.

### CODE REVIEW

File: pkg/kmodule/kmodule_linux_test.go [hunk 1]
```diff
@@ -32,3 +34,38 @@ func TestGenLoadedMods(t *testing.T) {
...
+	c := exec.Command("xz", "-f", n)
+	o, err := c.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("%s: %v", string(o), err)
+	}
+	return nil
+}
+
+// Not sure how to do what needs doing here.
+func testXZModules(t *testing.T) {
```
Comment (rjoleary) [inline pkg/kmodule/kmodule_linux_test.go]: Add a TODO saying this test is skipped.

File: pkg/kmodule/kmodule_linux.go [hunk 2]
```diff
@@ -237,31 +216,72 @@ func loadDeps(path string, m depMap, opts ProbeOpts) error {
...
 	if opts.DryRunCB != nil {
 		opts.DryRunCB(path)
 		return nil
 	}
 
 	f, err := os.Open(path)
-	if err != nil {
-		return err
-	}
 	defer f.Close()
```
Comment (rjoleary) [inline pkg/kmodule/kmodule_linux.go]: If `f == nil, won't this cause a nil panic?

File: pkg/kmodule/kmodule_linux.go [hunk 3]
```diff
@@ -237,31 +216,72 @@ func loadDeps(path string, m depMap, opts ProbeOpts) error {
...
+		}
+	}
+	if flags != 0 {
+		return fmt.Errorf("Can not have non-zero flags (%#x) with init_module", flags)
+	}
 
-	if err := FileInit(f, modParams, 0); err != nil && err != unix.EEXIST {
-		return err
+	// we're here; if the file open failed, try an xz file.
+	var r = io.Reader(f)
```
Comment (rjoleary) [inline pkg/kmodule/kmodule_linux.go]: If `f == nil`, won't this be an issue? Maybe put `r = io.Reader(f)` in an else clause.

File: pkg/kmodule/kmodule_linux.go [hunk 4]
```diff
@@ -237,31 +216,72 @@ func loadDeps(path string, m depMap, opts ProbeOpts) error {
...
+	if f == nil {
+		f, err := os.Open(path + ".xz")
+		if err != nil {
+			return fmt.Errorf("Neither %v or %v.xz can be opened", path, path)
```
Comment (rjoleary) [inline pkg/kmodule/kmodule_linux.go]: Also include error message in print.

File: pkg/kmodule/kmodule_linux.go [hunk 5]
```diff
@@ -237,31 +216,72 @@ func loadDeps(path string, m depMap, opts ProbeOpts) error {
...
+	if flags != 0 {
+		return fmt.Errorf("Can not have non-zero flags (%#x) with init_module", flags)
+	}
 
-	if err := FileInit(f, modParams, 0); err != nil && err != unix.EEXIST {
-		return err
+	// we're here; if the file open failed, try an xz file.
+	var r = io.Reader(f)
+	if f == nil {
+		f, err := os.Open(path + ".xz")
```
Comment (rjoleary) [inline pkg/kmodule/kmodule_linux.go]: Don't you need to `defer f.Close()` this file as well?

File: pkg/kmodule/kmodule_linux.go [hunk 6]
```diff
@@ -237,31 +216,72 @@ func loadDeps(path string, m depMap, opts ProbeOpts) error {
...
+	if err := LoadModule(path, "", 0, opts); err != nil {
 		return err
 	}
 	m[path].state = loaded
 
 	return nil
 }
 
-func loadModule(path, modParams string, opts ProbeOpts) error {
+// loadModule loads a file as a module. Because of all the various hacks -- er,
```
[the remark this answers is not in the document]
Comment (rminnich) [reply pkg/kmodule/kmodule_linux.go]: done

File: pkg/kmodule/kmodule_linux.go (hunk 2 above)
Comment (rminnich) [reply pkg/kmodule/kmodule_linux.go]: if err is nil, then f by definition is not nil? I don't know why this shows as deleted , it's not deleted in mine.

**Comment by rminnich** [comment]
I don't even have a way to test this. Somebody else can do it.

**Comment by srcUp** [comment]
@rminnich 
Please note that this issue only includes support for insmod, not for modprobe and depmod.
I hope we are tracking modprobe and depmod changes somewhere...

**Pull request closed without merging**